docs: improve interface doc (#9353)
dagger/dagger: docs/current_docs/api/interfaces.mdx

Note that you must also provide argument names, since they directly translate
to the GraphQL field argument names.

</TabItem>

<TabItem value="TypeScript">

The TypeScript SDK supports interfaces, which allow you to define set of functions
that a object must implement to be considered a valid instance of that interface.
That way, your module can accept arbitrary values from other modules without being
tightly coupled to the concrete type underneath.

To use an interface, use the TypeScript `interface` keyword and use it
as a function argument:

Here is an example of the definition of an interface `Fooer` with a single function `foo`:

```ts file=./snippets/interfaces/declarations/typescript/index.ts
```

Functions defined in interface definitions must match the client-side API
signature style:

> - Always define async function in interface (wrap the return type in a `Promise<T>`).

Reviewer: ```suggestion
- Always define `async` functions in interfaces (wrap the return type in a `Promise<T>`).
```

> - You can declare it as a method signature (e.g., `foo(): Promise<string>`) or a property signature (e.g., `foo: () => Promise<string>`).

Reviewer: ```suggestion
- Declare it as a method signature (e.g., `foo(): Promise<string>`) or a property signature (e.g., `foo: () => Promise<string>`).
```

`it` is ambiguous above, can you specify what `it` refers to?

Author: It's the function of the interface

- Parameters must be properly named since they directly translate to the GraphQL field argument names.

</TabItem>

</Tabs>

## Implementation

Here is an example of a module `Example` that implements the `Fooer` interface:

<Tabs groupId="language">
<TabItem value="Go">

```go file=./snippets/interfaces/implementations/go/main.go
```

</TabItem>
<TabItem value="TypeScript">

```ts file=./snippets/interfaces/implementations/typescript/index.ts
```

</TabItem>
</Tabs>

## Usage

Any object module that implements the interface method can be passed as an argument to the function
that uses the interface.

> Dagger automatically detects if an object coming for the module itself or one of its dependencies implements

Reviewer: ```suggestion
Dagger automatically detects if an object coming from the module itself or one of its dependencies implements
```

> an interface defined in its module or its dependencies. 

Reviewer: ```suggestion
an interface defined in the module or its dependencies. 
```

> If so, it will add new conversion function to the object that implement the interface

Reviewer: ```suggestion
If so, it will add new conversion functions to the object that implement the interface
```

so it can be passed as argument.

Here is an example of a module that uses the `Example` module defined above and

> pass it as argument of the `foo` function of the `MyModule` object:

Reviewer: ```suggestion
passes it as argument to the `foo` function of the `MyModule` object:
```
